libraries/aws/provisioning/src/aws_iot_provisioning_api.c

```diff
@@ -517,6 +533,195 @@ AwsIotProvisioningError_t AwsIotProvisioning_CreateKeysAndCertificate( IotMqttCo
+                                                                                 pPayloadBuffer,
+                                                                                 &payloadSize ) == false )
+        {
+            IotLogError( "Failed to serialize PUBLISH payload in buffer: Operation={%s}",
+                         CREATE_CERT_FROM_CSR_OPERATION_LOG );
+            status = AWS_IOT_PROVISIONING_INTERNAL_FAILURE;
+        }
+    }
+
+    if( status == AWS_IOT_PROVISIONING_SUCCESS )
+    {
+        publishInfo.qos = operationQos;
+        publishInfo.pPayload = pPayloadBuffer;
+        publishInfo.payloadLength = payloadSize;
+
+        /* Set the operation topic name. */
+        publishInfo.pTopicName = PROVISIONING_CREATE_CERT_FROM_CSR_REQUEST_TOPIC;
+        publishInfo.topicNameLength = PROVISIONING_CREATE_CERT_FROM_CSR_REQUEST_TOPIC_LENGTH;
+
+        IotLogDebug( "About to send request to server. Topic={%.*s}, Operation={%s}",
+                     publishInfo.topicNameLength,
+                     publishInfo.pTopicName,
+                     CREATE_CERT_FROM_CSR_OPERATION_LOG );
+
+        /* Publish to the Provisioning topic name. */
+        mqttOpResult = IotMqtt_PublishSync( connection,
+                                            &publishInfo,
+                                            0,
+                                            _AwsIotProvisioningMqttTimeoutMs );
+
+        if( mqttOpResult != IOT_MQTT_SUCCESS )
+        {
+            IotLogError( "Failed to publish to request topic: "
+                         "Topic={%.*s}, Operation={%s}, MQTTError={%s}",
+                         publishInfo.topicNameLength,
+                         publishInfo.pTopicName,
+                         CREATE_KEYS_AND_CERTIFICATE_OPERATION_LOG,
+                         IotMqtt_strerror( mqttOpResult ) );
+            status = AWS_IOT_PROVISIONING_MQTT_ERROR;
+        }
+    }
+
+    if( status == AWS_IOT_PROVISIONING_SUCCESS )
+    {
+        IotLogDebug( "Published to request topic: Operation={%s}",
+                     CREATE_CERT_FROM_CSR_OPERATION_LOG );
+
+        /* Wait for response from server using the given timeout period. */
+        status = _timedWaitForServerResponse( timeoutMs );
+
+        if( status == AWS_IOT_PROVISIONING_TIMEOUT )
+        {
+            IotLogDebug( "Operation timed out waiting for server response: Operation={%s}",
+                         CREATE_CERT_FROM_CSR_OPERATION_LOG );
+        }
+
+        /* Unsubscribe from the MQTT response topics only if subscription to those topics was successful. */
+        AwsIot_ModifySubscriptions( IotMqtt_UnsubscribeSync,
+                                    &responseSubscription );
+
+        AwsIotProvisioning_FreePayload( pPayloadBuffer );
+    }
+
+    IotLogInfo( "Operation is complete: Operation={%s}",
+                CREATE_CERT_FROM_CSR_OPERATION_LOG );
+
+    /* Reset the active operation */
+    _resetActiveOperationData();
```

Comment:
Is this global state? If so, this is an example of something that could be refactered into a caller-maintained context handle.

Reply:
But the state is internal to the functioning of the library...Do we want to expose the internal library state details? (Or is your suggestion that the library implementation should be state-less to make it _light-weight_ ? )

Reply:
Good point. For internal state, the caller context could either be entirely opaque, or be a public struct that includes a pointer to the opaque data.

Reply:
There is no exposure of the internal state in the APIs, it is completely encapsulated within the library implementation..

Reply:
Understood. It should stay that way. But we could eliminate the lock by having the caller hold our state in caller-supplied memory, and pass that state to us with each API call. A "handle" parameter is how that's frequently done.